### ISSUE TITLE
Type changes from MArray to SArray upon using broadcast (vectorization)

### ISSUE THREAD

**Issue by eeshan9815** [JuliaArrays/StaticArrays.jl#403]
```
julia> a = @MMatrix randn(2, 2)
2×2 StaticArrays.MArray{Tuple{2,2},Float64,2,4}:
 1.16721   -1.01285
 0.221507   0.00572563
```
```
julia> sa = sin.(a)
2×2 StaticArrays.SArray{Tuple{2,2},Float64,2,4}:
 0.919658  -0.848346
 0.2197     0.0057256
```
Why is `sa` of the type `SArray` whereas `a` is of the type `MArray`?

**Comment by tkoolen** [comment]
(Note: just a contributor). I think the current convention is to avoid allocation whenever possible. 
There's something to be said for either option, see also https://github.com/JuliaArrays/StaticArrays.jl/blob/8e36860a998c9db8452a5b868a43c184095cab5a/src/abstractarray.jl#L60-L61 but I think the current behavior would be frequently desirable (what with StaticArrays being all about getting the best performance possible).

Note that you can always do

```julia
julia> a = @MMatrix randn(2, 2)
2×2 StaticArrays.MArray{Tuple{2,2},Float64,2,4}:                                                                                                                                                                                                                                
 -0.266148  -0.146905                                                                                                                                                                                                                                                           
  0.67229    1.57561                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                
julia> sa = MMatrix{2, 2, Float64}()
2×2 StaticArrays.MArray{Tuple{2,2},Float64,2,4}:                                                                                                                                                                                                                                
 1.4822e-323  6.9015e-310                                                                                                                                                                                                                                                       
 6.9015e-310  0.0        

julia> sa .= sin.(a)
2×2 StaticArrays.MArray{Tuple{2,2},Float64,2,4}:
 -0.263017  -0.146378
  0.622779   0.999988
```

There would not be an easy workaround like the one above if an `MArray` were returned by default and you wanted an `SArray` without allocating.

**Comment by andyferris** [comment]
See also #327

**Comment by c42f** [comment]
Was solved in #536, which I'm now fairly confident was the right choice :-)
